### PR TITLE
 Bug 1623951 - convert hash urls and create a single entry point

### PR DESCRIPTION
**Pull request by sarah-clements**
This pr is finally done! Sorry, it's massive... I've had to make a lot of changes to the job-view logic and tests due to the routing changes. I've added the connected-react-router library so that redux dispatchers have access to the history object. Made routing changes much easier. A lot has likely changed since I checked this in as a WIP... let me know if you'd prefer for more of the commits to be squashed.

I've had one of the sheriffs test this pr on prototype but I'll ask for it to be tested again once on stage and also ask the perf sheriffs to do so.

### CODE REVIEW

File: ui/job-view/pushes/PushList.jsx
Comment (sarah-clements) [inline ui/job-view/pushes/PushList.jsx:218]: I hit an issue with this because the `fetchNextPushes` prop becauses it calls `addPushes` which uses `replaceLocation`. That way of updating params doesn't hit the `componentDidUpdate` method in job-view App that would trigger an update of FilterModel, which is involved in whether the filters are showing.

File: ui/job-view/pushes/PushList.jsx
Comment (sarah-clements) [inline ui/job-view/pushes/PushList.jsx:77]: Copied from the `reloadOnChangeParameters` with `onchange` removed because it calls `fetchPushes` again after `this.getNextPushes` is called because it hits `componentDidUpdate`, overwriting those results.

File: ui/job-view/pushes/PushList.jsx
Comment (camd) [reply ui/job-view/pushes/PushList.jsx:218]: It looks like it's only updating `fromChange` **before** the fetch of next pushes is complete.  So the range will be off.  It will need to update that param after the fetch is complete.

It also appears that changing the URL by hand causes a page reload, even if you're just modifying the `resultStatus` param.  This was one of the advantages of the hash URLs, since that gave us the ability to decide when to reload the page.  Is there a way to control that with `history'?  I know that you can listen for `history` changes with:

    history.listen((location) => {
        console.log(`You changed the page to: ${location.pathname}`)
    })

But this is the opposite issue of too much page reloading.  Perhaps `react-router-dom` has a way to control that.  Or perhaps we want to actually still use some params after the `#` like flitering to prevent page reload?  

I wonder if you may have to use a listener on `history` like:

    history.listen((location) => {
            console.log(`You changed the page to: ${location.pathname}`)
        })

A big part of using those event listeners was handling when someone changed the URL by hand.  Like changing the `resultStatus` property in this branch will reload the page.

**Comment by sarah-clements** [comment]
> I didn't get a ton of time to look at this, but I chimed in a bit. I know there is quite a bit of hoop-jumping that was done in Treeherder to update some things but not others. I tried to be very conscious of minimizing re-rendering as params changed. And this might be challenging to do without some params being after a `#` and using event listeners. That being said, we definitely don't need EVERYTHING after a hash. You could make the call on things like `repo`, `revision` and other range type items being before a hash and filtering after. Or just say that, if you manually tweak the URL, expect a page reload. I'm not sure how much impact that might have on folks like the sheriffs.

I've looked into it and there doesn't seem to be a way to prevent the browser from reloading when the url is manually changed without using hashes. As far as I can tell, you can't use the Browser router with hashes, only the Hash router and it'd probably be less confusing to do one or the other, not a combination with the query strings. 

I asked the sheriffs in the treeherder channel about it, and they only occasionally modify the url manually for date ranges. There might be a few users like jmaher who manually update query params more often but I don't know that this should be a blocker to implementing. So maybe we should go with "if you manually tweak the url, expect a page reload cause browsers". ¯\_(ツ)_/¯

File: ui/job-view/pushes/PushList.jsx
Comment (sarah-clements) [reply ui/job-view/pushes/PushList.jsx:218]: yeah, I was really doing it there so we have access to the location props and it was a bit confusing to use two different ways of updating the location. If you do it on the window object directly - which is how it's mostly done in the jobs view probably for the historical reasons you mentioned - it won't trigger a component life cycle update. But I'll take another look at this with fresh eyes :)

File: treeherder/config/settings.py
Comment (sarah-clements) [inline treeherder/config/settings.py:426]: For when the server is looking for the index file to return for any of our routes.

File: ui/App.jsx
Comment (sarah-clements) [inline ui/App.jsx:23]: Lazy loading of assets will keep our initial asset entry size about the same as what it was with the multiple entry points.

**Comment by camd** [comment]
I found a bug somewhere in the "fetch next pushes" functionality.  If you click on a single push, so we now have a `revision=<sha>` param, then click the get next `10` button, you should get 10 more pushes.  But you only get 9.  The `tochange` param is set correctly at the same sha that was formerly `revision`.  And the `fromchange` is correctly set to the last push fetched when you fetched 10.  The bug is that we're only seeing the next 9 pushes.

So you have 1 push, fetch 10 more and should have 11, but you have 10.  (What was that Joe's email said about off-by-one errors?)

Somewhere we are chopping off the last push we just fetched.  I remember we used to fetch 11 pushes and chop the last one off in case there was a range issue.  But I can see that when we click the `10` we make 2 calls for the `/push/?` endpoint.

For example, start here and hit next 10: http://localhost:5000/jobs?repo=autoland&resultStatus=testfailed%2Cbusted%2Cexception%2Crunnable&revision=6322b86e1df42a2d468a134d6580b4a5db746a9f

**Comment by camd** [comment]
We get 2 calls to the endpoint: 

http://localhost:5000/api/project/autoland/push/?full=true&count=11&push_timestamp__lte=1601991611
http://localhost:5000/api/project/autoland/push/?full=true&count=10&tochange=6322b86e1df42a2d468a134d6580b4a5db746a9f

**Comment by camd** [comment]
There's a lot here.  I likely won't complete reviewing this today.  I'll finish it tomorrow.

**Comment by sarah-clements** [comment]
> I found a bug somewhere in the "fetch next pushes" functionality. If you click on a single push, so we now have a `revision=<sha>` param, then click the get next `10` button, you should get 10 more pushes. But you only get 9. The `tochange` param is set correctly at the same sha that was formerly `revision`. And the `fromchange` is correctly set to the last push fetched when you fetched 10. The bug is that we're only seeing the next 9 pushes.
> 
> So you have 1 push, fetch 10 more and should have 11, but you have 10. (What was that Joe's email said about off-by-one errors?)
> 
> Somewhere we are chopping off the last push we just fetched. I remember we used to fetch 11 pushes and chop the last one off in case there was a range issue. But I can see that when we click the `10` we make 2 calls for the `/push/?` endpoint.
> 
> For example, start here and hit next 10: http://localhost:5000/jobs?repo=autoland&resultStatus=testfailed%2Cbusted%2Cexception%2Crunnable&revision=6322b86e1df42a2d468a134d6580b4a5db746a9f

This happened because of the same issue that snagged me in the beginning. If you update the `tochange` param via react-router instead of the window location object in `fetchNextPushes`, which I had to change because of another bug, then it gets hit in PushLists `componentDidUpdate` method (that's why you saw two requests). I've changed a bit of how that logic happens and found another place that needs to be changed. Should have it ready early Friday for review.

**Comment by camd** [comment]
I'm almost done reviewing this.  79/83 files...  I'll be sure to finish it tomorrow.

**Comment by camd** [comment]
I think this needs a rebase as well.  I merged the "mark investigated" PR from Suyash

File: ui/job-view/redux/stores/selectedJob.js
Comment (camd) [inline ui/job-view/redux/stores/selectedJob.js:243]: I'm curious how this will respond on prototype.  It's sluggish when switching between jobs using `n` and `p` as well as arrow right and left on my machine.  But the DB queries are much slower.  So it may just be that.  Passing in `updateJobDetails` here was an optimization.  I suppose we will find out when we get the Sheriffs to test-drive this before pushing to prod.  :)

**Review by camd — APPROVED** [review]
Overall, this looks great.  I'm pushing this branch to prototype now to do more testing.  I'll leave it up to you if you'd like the sheriffs to test on prototype or on stage after merge and just fix any issues in follow-ups.  I had a couple small things other than the `n` and `p` change in the `redux` store for `selectedJob`.  But this is a great upgrade!

File: ui/job-view/App.jsx
Comment (sarah-clements) [reply ui/job-view/App.jsx:76]: yes :)

File: ui/job-view/App.jsx
Comment (sarah-clements) [reply ui/job-view/App.jsx:208]: Sure.

**Comment by camd** [comment]
Heh, I found a small cosmetic bug.  Extraneous paren:

Screen Shot 2020-10-09 at 10 23 06 AM

File: ui/job-view/headerbars/SecondaryNavBar.jsx
Comment (sarah-clements) [reply ui/job-view/headerbars/SecondaryNavBar.jsx:130]: That is a good point. I think `pushRoute` makes sense; I'll update it where its used.

**Pull request merged**
